### PR TITLE
[Merged by Bors] - feat(Algebra/Polynomial/RingDivision): Add and simplify some lemmas

### PR DESCRIPTION
**Pull request by grhkm21**
Add `natDegree_pos_of_monic_of_ne_one` lemma which has a rather simple proof, and deduce the other lemmas from that.

---

I can't find a way to make `natDegree_pos_of_monic_of_ne_one` into a term mode proof, so more experienced Lean users, please teach me :)

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by adomani — COMMENTED** [review]
I know that the `a` symbol is already used in existing code, but I still think that recycling the available variables is better!

File: Mathlib/Algebra/Polynomial/RingDivision.lean [hunk 1]
```diff
@@ -345,6 +345,15 @@ theorem Monic.C_dvd_iff_isUnit {p : R[X]} (hp : Monic p) {a : R} :
       hp.coeff_natDegree ▸ (C_dvd_iff_dvd_coeff _ _).mp h p.natDegree,
    fun ha => (ha.map C).dvd⟩
 
+theorem natDegree_pos_of_monic_of_ne_one {a : R[X]} (hu : a ≠ 1) (ha : Monic a) :
+    0 < natDegree a := by
+  by_contra! ha'
+  exact hu <| eq_one_of_monic_natDegree_zero ha (by omega)
+
+theorem degree_pos_of_monic_of_ne_one {a : R[X]} (hu : a ≠ 1) (ha : Monic a) :
```
Comment (YaelDillies) [inline Mathlib/Algebra/Polynomial/RingDivision.lean:353]:
```suggestion
theorem Monic.degree_pos {a : R[X]} (hu : a ≠ 1) (ha : Monic a) :
```
or
```suggestion
theorem Monic.degree_pos_of_ne_one {a : R[X]} (hu : a ≠ 1) (ha : Monic a) :
```
or maybe even change the statement to
```
theorem Monic.degree_pos {a : R[X]} (hu : a ≠ 1) (ha : Monic a) :
```

Comment (YaelDillies) [reply Mathlib/Algebra/Polynomial/RingDivision.lean]: Whoop sorry I wrote nonsense at the end of my message. I meant maybe even change the statement to
```
theorem Monic.degree_pos {a : R[X]} (ha : Monic a) : 0 < degree a ↔  a ≠ 1
```

File: Mathlib/Algebra/Polynomial/RingDivision.lean [hunk 2]
```diff
@@ -337,33 +338,42 @@ end Ring
...
-theorem degree_pos_of_monic_of_not_isUnit {a : R[X]} (hu : ¬ IsUnit a) (ha : Monic a) :
-    0 < degree a :=
-  degree_pos_of_not_isUnit_of_dvd_monic hu dvd_rfl ha
-
-theorem natDegree_pos_of_monic_of_not_isUnit {a : R[X]} (hu : ¬ IsUnit a) (ha : Monic a) :
-    0 < natDegree a :=
-  natDegree_pos_iff_degree_pos.mpr <| degree_pos_of_monic_of_not_isUnit hu ha
+theorem Monic.natDegree_pos_of_ne_one (hu : p ≠ 1) : 0 < natDegree p := by
+  by_contra! hp'
+  exact hu <| eq_one_of_monic_natDegree_zero hp (by omega)
```
Comment (YaelDillies) [inline Mathlib/Algebra/Polynomial/RingDivision.lean:354]: I think you can golf this by first making `eq_one_of_monic_natDegree_zero` an iff (which you could rename to `Monic.natDegree_eq_zero`, which you can itself prove by replacing the use of `eq_C_of_natDegree_eq_zero` by `eq_C_coeff_zero_iff_natDegree_eq_zero` in the existing proof.

**Comment by grhkm21** [comment]
I have proven that now, I don't think it's the proof you were thinking of, but at least it works.

Just wondering, are there any policies on backwards compatability? For example, if I want to remove some lemmas, what do I have to do? I saw some `@[deprecated]` around, but don';t know how to use them. Is this written somewhere?

File: Mathlib/Algebra/Polynomial/RingDivision.lean [hunk 3]
```diff
@@ -349,12 +349,8 @@ theorem Monic.C_dvd_iff_isUnit {a : R} : C a ∣ p ↔ IsUnit a :=
       hp.coeff_natDegree ▸ (C_dvd_iff_dvd_coeff _ _).mp h p.natDegree,
    fun ha => (ha.map C).dvd⟩
 
-theorem Monic.natDegree_pos_of_ne_one (hu : p ≠ 1) : 0 < natDegree p := by
-  by_contra! hp'
-  exact hu <| eq_one_of_monic_natDegree_zero hp (by omega)
-
 theorem Monic.natDegree_pos : 0 < natDegree p ↔ p ≠ 1 :=
-  ⟨fun hu ↦ by contrapose! hu; rw [hu]; compute_degree!, fun hu ↦ hp.natDegree_pos_of_ne_one hu⟩
+  Nat.pos_iff_ne_zero.trans (Monic.natDegree_eq_zero hp).not
```
Comment (YaelDillies) [inline Mathlib/Algebra/Polynomial/RingDivision.lean:353]:
```suggestion
  Nat.pos_iff_ne_zero.trans hp.natDegree_eq_zero.not
```

**Comment by YaelDillies** [comment]
> Just wondering, are there any policies on backwards compatibility? For example, if I want to remove some lemmas, what do I have to do? I saw some `@[deprecated]` around, but don't know how to use them. Is this written somewhere?

You should do `@[deprecated new_lemma (since := "2024-06-18")] lemma old_lemma ...`

**Comment by YaelDillies** [comment]
Note that you have a shake issue

**Review by YaelDillies — APPROVED** [review]
Looks good now

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

File: Mathlib/Algebra/Polynomial/Degree/Definitions.lean [hunk 4]
```diff
@@ -1215,6 +1215,9 @@ theorem eq_one_of_monic_natDegree_zero (hf : p.Monic) (hfd : p.natDegree = 0) :
   rw [Monic.def, leadingCoeff, hfd] at hf
   rw [eq_C_of_natDegree_eq_zero hfd, hf, map_one]
 
+theorem Monic.natDegree_eq_zero (hf : p.Monic) : p.natDegree = 0 ↔ p = 1 :=
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Polynomial/Degree/Definitions.lean:1218]:
```suggestion
theorem Monic.natDegree_eq_zero_iff (hf : p.Monic) : p.natDegree = 0 ↔ p = 1 :=
```

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/Algebra/Polynomial/Degree/Definitions.lean (hunk 4 above)
Comment (YaelDillies) [reply Mathlib/Algebra/Polynomial/Degree/Definitions.lean:1218]: Actually here I think the `_iff` is pretty clearly implied by the rest

**Pull request closed without merging**